Review "Links" documentation (#501)
grafana/writers-toolkit: docs/sources/write/links/_index.md

The source is reused as described in [Reuse directories of content with Hugo mounts](https://grafana.com/docs/writers-toolkit/write/reuse-content/reuse-directories/).
For more information and examples, refer to [`docs/reference` shortcode](https://grafana.com/docs/writers-toolkit/write/shortcodes/#docsreference).

## Link to `grafana.com` pages

Use a full URL.

{{< admonition type="note" >}}
The `doc-validator` linter doesn't check links that use full URLs.
{{< /admonition >}}

If you are linking to versioned documentation, use a full URL with version substitution syntax instead of the version path element.

> For example, in Grafana, use `<GRAFANA_VERSION>` instead of `latest` in the URL `https://grafana.com/docs/grafana/latest/`.

Reviewer: This kind of caught me out the first time I read it.. .I read it as just use `<GRAFANA_VERSION>` as the URL (despite knowing the correct way).  I don't have a suggestion here... just noting that it's one of those "read carefully" sentences.

Author: Mm, thanks for raising this. I think it is more clear in https://grafana.com/docs/writers-toolkit/write/shortcodes/#about-version-substitution:~:text=As%20a%20convention%2C%20use%20the%20name%20of%20the%20target%20project%20all%20upper%2Dcase.%20For%20example%2C%20grafana%20becomes%20GRAFANA%2C%20grafana%2Dcloud%20becomes%20GRAFANA%20CLOUD.

Perhaps that should be here and the "About ..." should more on the behavior under the hood.

> Version substitution replaces the `<SOMETHING_VERSION>` syntax with the version inferred from the current page.

Reviewer 2: ```suggestion
When the documentation is published to the website, version substitution replaces the `<SOMETHING_VERSION>` syntax with the version inferred from the current branch for the documentation page.
```

Author: It's not really true that version substitution happens during publishing. It happens when the link is rendered. Perhaps:

```suggestion
When Hugo renders links with version substitution, it replaces the `<SOMETHING_VERSION>` syntax with the version inferred from the current page.
```

> When Hugo renders links with version substitution, it replaces the `<SOMETHING_VERSION>` syntax with the version inferred from the current page.

Reviewer 2: As far as I know, there is zero data in a markdown topic to indicate the version.  If I'm understanding correctly how this  magic works, isn't it the BRANCH that the page is being published to that determines the version?

Reviewer: Agent uses this feature to link to specific versions of external and internal documentation... and sets the referenced version as a `cascade`.

Author: It comes from the relative permalink for the page rather than from the source file. Hugo knows the relative permalink for the page at build time and that contains a version that we extract as part of the link.

The page `docs/sources/loki/_index.md` from the Loki repository is synced to the `content/docs/loki/<VERSION>/` directory in the website repository and at Hugo build time the page has relative permalink `/docs/loki/<VERSION>/`.

It is from that relative permalink that we can extract the version and is what is intended by the phrase "inferred from the page".

If you think in terms of pages rather than source files, it perhaps is simpler.
If you are on the page `/docs/grafana/latest/dashboards/` version substitution replaces the `<GRAFANA_VERSION>` in links with `latest` from the page URL.

All of this is the default behavior. Front matter overrides this if it is set.

Reviewer 2: Could we add this information to the Writers Toolkit?  Because without understanding what magic Hugo is working behind the scenes, the way this is currently written implies that the version is somehow extracted from the markdown file for a page.  Which doesn't seem possible (and isn't, as you've described what's really happening).

Author: I think that's in https://grafana.com/docs/writers-toolkit/write/shortcodes/#about-version-substitution. What about linking to that?

Initially the advice from Isabel for this page was to focus on what to do rather than how stuff works.
Unfortunately, it feels to me that depending on who looks at the page, we endlessly oscillate back and forth from explaining everything to just telling people what to do.

```suggestion
When Hugo renders links with version substitution, it replaces the `<SOMETHING_VERSION>` syntax with the version inferred from the current page.

To understand the behavior in more detail, refer to [About version substitution](https://grafana.com/docs/writers-toolkit/write/shortcodes/#about-version-substitution)
```

Reviewer 2: > we endlessly oscillate back and forth from explaining everything to just telling people what to do.

I admit, I am the Elephant's Child, always asking "Why?"  I don't know that the About Version Substitution page makes any more sense to me.  

Maybe we should just revise this line to read "When Hugo renders links with version substitution, it replaces the `<SOMETHING_VERSION>` syntax with the correct version."  and not say anything about how the magic happens.

Reviewer: I'm a big supporter of NOT mentioning the magic behind the curtain in the user/contributor facing documentation. Save the details for a child page or a "how it all works" doc set for the maintainers. Here, just saying "it does a thing when you do this" is probably sufficient... and either link out to the how it works, or keep the how it works in a design doc set somewhere else?

> It's necessary for full URLs to link to the correct version of documentation without requiring the writer to update the version for each release.

Reviewer 2: ```suggestion
You use full URLs with the substitution variable so that links will resolve to the correct version of documentation without requiring the writer to update the version for each release.
```
